src/placement/jump_map.c

```diff
@@ -365,28 +367,35 @@ get_target(struct pool_domain *curr_dom, struct pool_target **target,
 			end_tgt = start_tgt + (num_doms - 1);
 
 			range_set = isset_range(tgts_used, start_tgt, end_tgt);
-			if (range_set)
-				clrbit_range(tgts_used, start_tgt, end_tgt);
-
-			do {
-				/*
-				 * Must crc key because jump consistent hash
-				 * requires an even distribution or it will
-				 * not work
-				 */
-				obj_key = crc(obj_key, fail_num++);
-
-				/* Get target for shard */
-				selected_dom = jump_consistent_hash(obj_key,
-								    num_doms);
+			if (range_set) {
+				/* Used up all targets in current domain,
+				 * let's try parent domain
+ 				 */
+				setbit(dom_occupied, curr_dom - root_pos);
+				D_ASSERT(top != -1);
+				curr_dom = dom_stack[top--];
+				continue;
+			}
 
-				/* Retrieve actual target using index */
-				*target = &curr_dom->do_targets[selected_dom];
+			/*
+			 * Must crc key because jump consistent hash
+			 * requires an even distribution or it will
+			 * not work
+			 */
+			obj_key = crc(obj_key, fail_num++);
 
-				/* Get target id to check if target used */
-				dom_id = (*target)->ta_comp.co_id;
+			/* Get target for shard */
+			selected_dom = jump_consistent_hash(obj_key, num_doms);
 
-			} while (isset(tgts_used, dom_id));
+			do {
+				selected_dom = selected_dom % num_doms;
+ 				/* Retrieve actual target using index */
+ 				*target = &curr_dom->do_targets[selected_dom];
+
+ 				/* Get target id to check if target used */
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs